mdbook-tera-backend/src/tera_renderer/custom_component.rs

```diff
@@ -0,0 +1,133 @@
+use serde::Deserialize;
+use std::cell::RefCell;
+use std::collections::{BTreeMap, HashMap};
+use std::path::{Path, PathBuf};
+
+use tera::Tera;
+
+use crate::{RendererError, Result};
+
+use super::RenderingContext;
+
+fn make_strip_prefix_function() -> impl tera::Function {
+    move |args: &HashMap<String, serde_json::value::Value>| -> tera::Result<tera::Value> {
+        let string = args
+            .get("s")
+            .ok_or_else(|| tera::Error::from(format!("No s argument provided")))?
+            .as_str()
+            .ok_or_else(|| tera::Error::from(format!("S has invalid type, expected string")))?;
+        let prefix = args
+            .get("prefix")
+            .ok_or_else(|| tera::Error::from(format!("No prefix argument provided")))?
+            .as_str()
+            .ok_or_else(|| {
+                tera::Error::from(format!("Prefix has invalid type, expected string"))
+            })?;
+        string
+            .strip_prefix(prefix)
+            .map(|s| tera::Value::String(s.to_owned()))
+            .ok_or_else(|| tera::Error::from(format!("Could not strip prefix")))
+    }
+}
+
+pub struct CustomComponent {
+    template: Tera,
+    name: String,
+    /// Used to generate unique ids for each component to prevent collisions in javascript with query selectors.
+    counter: RefCell<u32>,
```

Comment:
This code looks very custom to how you will use it in Comprehensive Rust?

Reply:
I think it's essential if you're adding a script with `getElementById` and you want to add the same component multiple times in the page, they will all have a different id thanks to `{{counter}}` so it's definitely something that people will want

Reply:
> same component

My concern is that you're inventing new vocabulary here — don't design "web components", just make a trivial system which runs Tera on the HMTL. The simplest thing which will let us solve the issues that motivate this effort.

If people want to call the same Tera macro multiple times, then they can set and maintain a counter, no?

Reply:
Maybe, I haven't had time to dive on all the features, so it might be possible to add a global variable like this but I don't know how